sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowEvalPythonExec.scala

```diff
@@ -170,64 +170,106 @@ class RebatchingRoundoffIterator(
   }
 }
 
+/** A queue that consumes an Iterator and provides peek and remove operations */
+trait IteratorBatchQueue {
+  def peekBatchNumRows(): Int
+
+  def remove(): ColumnarBatch
+}
+
 /**
- * A simple queue that holds the pending batches that need to line up with
- * and combined with batches coming back from python
+ * An iterator that will cache the batches in a queue after pulling in to support
+ * the IteratorBatchQueue.
+ * The cached batches will be used for later combination each time getting
+ * a result batch from the Python side.
+ * It will transform the input batch by "cacheConverter" and cache the result.
```

Comment:
I am really confused by this. So we modify the data twice. Why? We have a `queue` and `pending`. Batches from `queue` are returned by `remove`.  But pending comes from `next`? It feels like what you really want is a way to split an input iterator into two and then provide different operations that can be done on the two. The interface here is the confusing part for me. I would much rather see us have an object that will take the input and the two converters and return two iterators.

Also this is always used with the `CombiningIterator`. Should we just put all of that together and have a single API that takes an input iterator, and two separate functions that will process that input in different ways. It will return a single iterator that has combined the two results together after doing the processing?

Reply:
>Should we just put all of that together and have a single API that takes an input iterator, and two separate functions that will process that input in different ways.

No, `CombiningIterator` depends on the Python runner, it can not be an input to the Python runner. One of the data pipeline is used as the runner input.

Reply:
I removed this new class, and updated the original `BatchQueue` to support asking for a batch instead of waiting until one is inserted when the writer is writing a batch to Python.